src/GraphQL/Extensions/ObjectExtensions.cs

```diff
@@ -28,175 +28,268 @@ public static class ObjectExtensions
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.PublicFields | DynamicallyAccessedMemberTypes.PublicProperties)]
+            Type type,
+            IGraphType mappedType)
         {
             var inputGraphType = (mappedType is NonNullGraphType nonNullGraphType
                 ? nonNullGraphType.ResolvedType as IInputObjectGraphType
                 : mappedType as IInputObjectGraphType)
                 ?? throw new InvalidOperationException($"Graph type supplied is not an input object graph type.");
 
-            // Given Field("FirstName", x => x.FName) and key == "FirstName" returns "FName"
-            string GetPropertyName(string key, out FieldType field)
+            if (source == null)
+                throw new ArgumentNullException(nameof(source));
+
+            // if conversion from IDictionary<string, object> to desired type is registered then use it
+            if (ValueConverter.TryConvertTo(source, type, out object? result, typeof(IDictionary<string, object>)))
+                return result!;
+
+            var reflectionInfo = GetReflectionInformation(type, inputGraphType);
+
+            // build the constructor arguments
+            object?[] ctorArguments = reflectionInfo.CtorFields.Length == 0
+                ? Array.Empty<object>()
+                : new object[reflectionInfo.CtorFields.Length];
+
+            for (int i = 0; i < reflectionInfo.CtorFields.Length; ++i)
             {
-                // type may not contain mapping information
-                field = inputGraphType.GetField(key)
-                    ?? throw new InvalidOperationException($"Could not find field '{key}' on type '{inputGraphType}'.");
-                return field.GetMetadata(ComplexGraphType<object>.ORIGINAL_EXPRESSION_PROPERTY_NAME, key) ?? key;
+                var ctorField = reflectionInfo.CtorFields[i];
+                ctorArguments[i] = ctorField.Key != null
+                    ? GetPropertyValue(source.TryGetValue(ctorField.Key, out var value) ? value : null, ctorField.ParameterInfo.ParameterType, ctorField.GraphType!)
+                    : ctorField.ParameterInfo.DefaultValue;
             }
 
-            // Returns values (from source or defaults) that match constructor signature + used keys from source
-            (List<object?>?, List<string?>?) GetValuesAndUsedKeys(ParameterInfo[] parameters)
+            // construct the object
+            object obj;
+            try
             {
-                // parameterless constructors are the most common use case
-                if (parameters.Length == 0)
-                    return (_emptyValues, null);
-
-                // otherwise we have to iterate over the parameters - worse performance but this is rather rare case
-                List<object?>? values = null;
-                List<string?>? keys = null;
+                obj = reflectionInfo.Constructor.Invoke(ctorArguments);
```

Comment:
`Activator.CreateInstance` should be [faster](https://andrewlock.net/benchmarking-4-reflection-methods-for-calling-a-constructor-in-dotnet/).